feat: Rstudio provisioning with ALB

manikandan-thangavelu-rl opened the pull request:
Issue #, if available:

Description of changes:

- Added an additional step in provision workflow environment that checks for the flag NeedsALB in the CFT output and provisions an ALB if needed.
- Created a new service alb-service that contains the code related to ALB.
- Creating Route53 record with ALB DNS for RstudioV2
- Creating a listener rule for the ALB for RstudioV2
- Added scripts in bootstrap.sh and provision-rstudio.sh to create self signed ssl certificate and additional R packages.
- Made the code compatible with older Rstudio version and RstudioV2.
- Added a new template 'application-load-balancer' to the cfnTemplateService.
- Added permissions required to read cf-templates-* S3 bucket and DeploymentStore table for the backend lambda in CFT.

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?
- [x] Have you successfully tested with your changes locally?
- [x] If new dependencies have been added, have they been pinned to specific versions?

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
fix: adds 'stopped' filter for workspaces

SanketD92 committed:
chore: dependabot suggested package versions (#488)

nguyen102 committed:
docs: Update docs to specify Packer version 1.6.0 (#484)

ahl27 committed:
bulkUserAdd API bugfix (#490)

* fix: bug in openapi.yaml introduced in previous PR with unescaped asterisk causing malformed YAML file
* fix: adds validation checks and more informative error messages for bulk-add-users API call.
* fix: added unit and integration tests for changes

ahl27 committed:
fix: moved notification boxes to avoid blocking the top ribbon. (#483)

manikandan-thangavelu-rl committed:
Rstudio DEV Added ALB provisioing as a step in launch workflow

manikandan-thangavelu-rl committed:
Removed projectId from input payload

manikandan-thangavelu-rl committed:
Removed policy for elb

manikandan-thangavelu-rl committed:
Merge branch 'mainline' into rstudio-dev

manikandan-thangavelu-rl committed:
Merge pull request #1 from manikandan-thangavelu-rl/rstudio-dev

Rstudio dev to Developer mainline

rsmayda committed:
chore: Update UI to show 3.1 (#493)

SanketD92 committed:
chore: Update issue templates (#487)

* chore: Update issue templates

* Trigger notification

* fix: changes per review

Co-authored-by: Robert Smayda <smayda44@gmail.com>

nguyen102 committed:
fix: Redirect non admin users to "/" if they try to access "/users" (#489)

manikandan-thangavelu-rl committed:
Removed unwanted logs

SanketD92 committed:
chore: update PR checklist (#494)

manikandan-thangavelu-rl committed:
Modified ALB CFT

manikandan-thangavelu-rl committed:
Changed var naming convention, added comments to ignore temporary code changes

RLIndia committed:
Merge pull request #1 from manikandan-thangavelu-rl/mainline

feat: Rstudio provisioning with ALB

SanketD92 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -0,0 +1,298 @@
+    domainName: 'domainName',
+};
+
+class ALBService extends Service {
```
Missing unit tests.

manikandan-thangavelu-rl replied:
We are working on unit tests. We will be adding the unit tests in subsequent PRs

SanketD92 replied:
I'd recommend adding the unit tests in this PR itself, since the automated code coverage check would fail.

SanketD92 commented on addons/addon-base-raas/packages/base-raas-services/lib/plugins/env-provisioning-plugin.js:
```diff
@@ -384,12 +407,20 @@ async function rstudioCleanup(requestContext, updatedEnvironment, container) {
+        await environmentDnsService.deleteRecord('rstudio', updatedEnvironment.id, dnsName);
+      }
+      else if (connectionTypeValue.toLowerCase() === 'rstudiov2') {
+        // TODO: Add code to remove record for new Rstudio version
```
Will this be addressed in the current PR?

manikandan-thangavelu-rl replied:
This will be part of the terminate workflow. So we will address this in the respective PR raised for the termination workflow.

SanketD92 commented on addons/addon-base-raas/packages/base-raas-services/lib/plugins/env-provisioning-plugin.js:
```diff
@@ -384,12 +407,20 @@ async function rstudioCleanup(requestContext, updatedEnvironment, container) {
+        //Moved the code here to handle deletion of RstudioV1, This will be changed when termination
+        // code for RstudioV2 comes in. Please ignore this on code review
+        const environmentDnsService = await container.find('environmentDnsService');
+        await environmentDnsService.deleteRecord('rstudio', updatedEnvironment.id, dnsName);
```
Can this be configurable? What if users want to continue using the non-ALB version of RStudio?

manikandan-thangavelu-rl replied:
We will be having a condition check for Rstudio and RstudioV2 so that the user can still continue to use the existing Rstudio

manikandan-thangavelu-rl committed:
Added ALB permisison for SC launch constraint role, and changed ALB stack name to analysis-*

nguyen102 committed:
chore: Add GH action to check for dependency vulnerabilities  (#498)

manikandan-thangavelu-rl committed:
Removed unwanted comments

RLIndia committed:
Merge pull request #2 from manikandan-thangavelu-rl/rstudio-terminate

fix: Added ALB permission for SC launch constraint role

nguyen102 committed:
fix: react compilation error (#500)

shyogesh-sw committed:
docs: hyperlink corrections (#497)

* <docs: added hyperlinks in packer readme>

* Revert "<docs: added hyperlinks in packer readme>"

This reverts commit 501b7de9161296361c761d0d9a18d27cf73ae8aa.

* docs: corrected the hyperlinks

nguyen102 committed:
feat: study permissions only shown to Study Admin (#501)

manikandan-thangavelu-rl committed:
feat: terminate lauch dependencies before terminating peroduct

manikandan-thangavelu-rl committed:
Removed unwanted logs

rsmayda committed:
fix: add termination status for non-found workspaces (#502)

nguyen102 committed:
fix: Do not allow users to change root password (#503)

jn1119 commented on addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/application-load-balancer.cfn.yml:
```diff
@@ -0,0 +1,70 @@
+    Type: AWS::EC2::Subnet
+    Properties:
+      VpcId: !Ref VPC
+      CidrBlock: 10.0.32.0/19
```
Can you explain why CidrBlock is hardcoded to `10.0.32.0/10`

manikandan-thangavelu-rl replied:
This is based on the default value of the Subnet1 CIDR on the onboard account CFT

jn1119 commented on addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/application-load-balancer.cfn.yml:
```diff
@@ -0,0 +1,70 @@
+    Properties:
+      VpcId: !Ref VPC
+      CidrBlock: 10.0.32.0/19
+      AvailabilityZone: us-east-1b
```
Is it expected for region to be `us-east-1`. Solution can be deployed in any region, right?

manikandan-thangavelu-rl replied:
Updated the CFT to fetch the availability zone dynamically.

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -0,0 +1,314 @@
+    async findAwsAccountId(requestContext, projectId) {
+        const [
+            indexesService,
+            ProjectService,
```
Can you rename this to `projectService` for consistency

manikandan-thangavelu-rl replied:
We will make this change

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -0,0 +1,314 @@
+        const deploymentItem = await this.getAlbDetails(requestContext, resolvedVars.projectId);
+        const albRecord = JSON.parse(deploymentItem.value);
+        const listenerArn = albRecord.listenerArn;
+        const priority = albRecord.albDependentWorkspacesCount + 1;
```
Do we need to worry about effect concurrent requests will have on priority?

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -0,0 +1,314 @@
+    async increaseAlbDependentWorkspaceCount(requestContext, resolvedVars) {
+        const deploymentItem = await this.getAlbDetails(requestContext, resolvedVars.projectId);
+        var albRecord = JSON.parse(deploymentItem.value);
+        albRecord.albDependentWorkspacesCount = albRecord.albDependentWorkspacesCount + 1;
```
Do we need to worry about potentially the field not being correctly incremented in case of parallel requests?

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -0,0 +1,314 @@
+    async increaseAlbDependentWorkspaceCount(requestContext, resolvedVars) {
+        const deploymentItem = await this.getAlbDetails(requestContext, resolvedVars.projectId);
+        var albRecord = JSON.parse(deploymentItem.value);
+        albRecord.albDependentWorkspacesCount = albRecord.albDependentWorkspacesCount + 1;
```
Do we need to worry about concurrent requests changing the value of `albDependentWorkspacesCount` and the value not getting incremented correctly?

manikandan-thangavelu-rl replied:
The concurrent request issues will be resolved once the lock service is implemented. You can expect that in the next PR

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -0,0 +1,314 @@
+        const deploymentItem = await this.getAlbDetails(requestContext, resolvedVars.projectId);
+        const albRecord = JSON.parse(deploymentItem.value);
+        const listenerArn = albRecord.listenerArn;
+        const priority = albRecord.albDependentWorkspacesCount + 1;
```
Do we need to worry about concurrent requests where `priority` doesn't get updated correctly since the value of `albDependentWorkspacesCount` can be changed in parallel.

manikandan-thangavelu-rl replied:
The concurrent request issues will be resolved once the lock service is implemented. You can expect that in the next PR

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/plugins/env-provisioning-plugin.js:
```diff
@@ -220,6 +220,29 @@ async function updateEnvOnProvisioningSuccess({
     }
+    else if (connectionTypeValue.toLowerCase() === 'rstudiov2') {
+      const albService = await container.find('albService');
+      const deploymentItem = await albService.getAlbDetails(requestContext, resolvedVars.projectId);
+      const albExists = await albService.checkAlbExists(requestContext, resolvedVars.projectId);
```
Seems like as part of both the calls here we are calling `getAlbDetails`. Can we avoid that and do that just once.

manikandan-thangavelu-rl replied:
We will make this change

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/plugins/env-provisioning-plugin.js:
```diff
@@ -220,6 +220,29 @@ async function updateEnvOnProvisioningSuccess({
+      const environmentDnsService = await container.find('environmentDnsService');
+      await environmentDnsService.createRecord('rstudio', envId, dnsName);
+      //Create a listener rule
+      const ruleARN = await albService.createListenerRule('rstudio', requestContext, resolvedVars, targetGroupArn);
+      // Save the Rule ARN as an output in DB
+      const ruleRecord = {
+        Description: "ARN of the listener rule created by code",
+        OutputKey: "ListenerRuleARN",
+        OutputValue: ruleARN
+      };
+      outputs.push(ruleRecord);
+      await albService.increaseAlbDependentWorkspaceCount(requestContext, resolvedVars);
```
Seems like we need to put this code under a distributed lock so that we don't potentially miss an update. Take a look at [lock-service](https://github.com/awslabs/service-workbench-on-aws/blob/mainline/addons/addon-base/packages/services/lib/lock/lock-service.js) and [example code](https://github.com/awslabs/service-workbench-on-aws/blob/mainline/addons/addon-base-raas/packages/base-raas-services/lib/aws-accounts/aws-accounts-service.js#L91)

manikandan-thangavelu-rl replied:
The concurrent request issues will be resolved once the lock service is implemented. You can expect that in the next PR

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+    * @returns {Promise<>}
+    */
+    async provisionAlb(requestContext, resolvedVars, projectId, resolvedInputParams, maxAlbWorkspacesCount) {
+        const [ALBService] = await this.mustFindServices(['albService']);
```
rename to `albService` for consistency?

manikandan-thangavelu-rl replied:
We will make this change

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+            throw new Error(`Error provisioning environment. Reason: Maximum workspaces using ALB has reached`);
+        }
+        if (albExists) {
+            this.print({
+                msg: `ALB Already exists for the Account. Skipping ALB Creation`,
+            });
```
Should be a log?

manikandan-thangavelu-rl replied:
Yes. Added a log to debug if the ALB provisioning was needed. Can be removed if suggested otherwise.

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+                msg: `ALB Already exists for the Account. Skipping ALB Creation`,
+            });
+        } else {
+            this.print({
+                msg: `Workspace needs ALB. Provisioning an ALB.`,
+            });
```
Should be a log?

manikandan-thangavelu-rl replied:
Yes. Added a log to debug if the ALB provisioning was needed. Can be removed if suggested otherwise.

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+    */
+    async provisionAlb(requestContext, resolvedVars, projectId, resolvedInputParams, maxAlbWorkspacesCount) {
+        const [ALBService] = await this.mustFindServices(['albService']);
+        const count = await ALBService.albDependentWorkspacesCount(requestContext, projectId);
+        const albExists = await ALBService.checkAlbExists(requestContext, projectId);
+        if (count >= maxAlbWorkspacesCount) {
+            throw new Error(`Error provisioning environment. Reason: Maximum workspaces using ALB has reached`);
+        }
```
Don't think this entire condition is safe against parallel calls. Consider using lock-service here for distributed locks.

manikandan-thangavelu-rl replied:
The concurrent request issues will be resolved once the lock service is implemented. You can expect that in the next PR

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+            this.print({
+                msg: `Workspace needs ALB. Provisioning an ALB.`,
+            });
+            const stackInput = await ALBService.getStackCreationInput(requestContext, resolvedVars, resolvedInputParams, projectId);
+            //Storing Dependency type so the stack completion can be handled for different dependencies
+            this.state.setKey('DEPENDENCY_TYPE', 'ALB');
+            //Create Stack
+            return await this.deployStack(requestContext, resolvedVars, stackInput);
```
What if ALB doesn't exist but we have one in-flight. We want to avoid deploying this stack in case one is in progress, right?

manikandan-thangavelu-rl replied:
The concurrent request issues will be resolved once the lock service is implemented. You can expect that in the next PR

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+        this.state.setKey('STACK_ID', response.StackId);
+        return (
+            this.wait(5) // check every 5 seconds
+                // keep doing it for 1*1296000 seconds = 15 days
```
Any reason we need to do it for 15 days? Does it take that long?

manikandan-thangavelu-rl replied:
We have followed the same approach as the launch-product step where the actual product is launched. Please let us know if this needs to be changed

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -0,0 +1,433 @@
+    * @returns {Promise<*>}
+    */
+    async onFail(error) {
+        this.printError(error);
```
Where is this function defined? Can we log instead?

manikandan-thangavelu-rl replied:
We have followed the same approach as the launch-product step to print any errors. Please let us know if this needs to be changed.

RLIndia committed:
Merge pull request #3 from manikandan-thangavelu-rl/rstudio-terminate

feat: Terminate ALB step in termination workflow

Bingjiling committed:
chore: docs dependency fix (#505)

dependabot[bot] committed:
chore(deps): bump dns-packet from 1.3.1 to 1.3.4 in /docs (#507)

* chore(deps): bump dns-packet from 1.3.1 to 1.3.4 in /docs

Bumps [dns-packet](https://github.com/mafintosh/dns-packet) from 1.3.1 to 1.3.4.
- [Release notes](https://github.com/mafintosh/dns-packet/releases)
- [Changelog](https://github.com/mafintosh/dns-packet/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mafintosh/dns-packet/compare/v1.3.1...v1.3.4)

Signed-off-by: dependabot[bot] <support@github.com>

RLIndia committed:
Merge pull request #4 from RLOpenCatalyst/develop

Develop

manikandan-thangavelu-rl committed:
fix: incorporated PR comments on code style

RLIndia committed:
Merge pull request #6 from manikandan-thangavelu-rl/feat-rstudio-alb

fix: incorporated PR comments on code style

manikandan-thangavelu-rl committed:
Modified pnpm lock to fix PR issue

manikandan-thangavelu-rl committed:
fix static code analysis and linting issues

RLIndia committed:
Merge pull request #9 from manikandan-thangavelu-rl/feat-rstudio-alb

Feat rstudio alb

manikandan-thangavelu-rl committed:
Added unit test cases for alb-service

RLIndia committed:
Merge pull request #10 from manikandan-thangavelu-rl/feat-rstudio-alb

Added unit test cases for alb-service

nguyen102 committed:
feat: Add warning that internal authentication shouldn't be used in production (#506)

manikandan-thangavelu-rl committed:
Fetching availability zones dynamically in alb cft

RLIndia committed:
Merge pull request #11 from manikandan-thangavelu-rl/feat-rstudio-alb

Fetching availability zones dynamically in alb cft

nguyen102 committed:
feat: Encrypt s3 buckets for EMR log bucket and CICD Artifact bucket (#508)

nguyen102 committed:
chore: Disable EBS volume for storage gateway (#511)

Co-authored-by: Tim Nguyen <thingut@amazon.com>

nguyen102 committed:
chore: Add encryption to CICD SNS topic (#512)

Co-authored-by: Tim Nguyen <thingut@amazon.com>

jn1119 reviewed (commented):
For future reference- It would be helpful if merging prior commits from develop was something that is done in a separate change so that the reviewer can just focus on specific change being requested in the PR.

jn1119 commented on addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/application-load-balancer.cfn.yml:
```diff
@@ -0,0 +1,70 @@
+    Type: AWS::EC2::VPC::Id
+  Subnet1:
+    Description: The VPC subnet in which the EC2 instance will reside
+    Type: AWS::EC2::Subnet::Id
```
Is the description accurate here? I guess we should mention ALB here. What do you think?

manikandan-thangavelu-rl replied:
Made the change and updated the PR

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:
```diff
@@ -17,298 +17,327 @@ const _ = require('lodash');
+   * @returns {Promise<int>}
+   */
+  async albDependentWorkspacesCount(requestContext, projectId) {
+    const deploymentItem = await this.getAlbDetails(requestContext, projectId);
```
I am guessing that the comments related to locking aren't resolved yet?

manikandan-thangavelu-rl replied:
Comments related to locking are resolved and raised a separate PR.

nguyen102 committed:
fix: Allow sagemaker to have the proper IAM permission to autostop itself (#515)

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js:
```diff
@@ -0,0 +1,433 @@
+      });
+      service.audit = jest.fn();
+      // OPERATE
+      await service.increaseAlbDependentWorkspaceCount({}, 'test-id');
```
Shouldn't this verify if the count was increased successfully or not?

manikandan-thangavelu-rl replied:
Made the change and updated the PR

jn1119 commented on addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js:
```diff
@@ -0,0 +1,433 @@
+      });
+      service.audit = jest.fn();
+      // OPERATE
+      await service.decreaseAlbDependentWorkspaceCount({}, 'test-id');
```
Shouldn't this verify if the count was decreased successfully or not?

manikandan-thangavelu-rl replied:
Made the change and updated the PR

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -64,370 +64,386 @@ const successStatuses = ['CREATE_COMPLETE', 'DELETE_COMPLETE', 'UPDATE_COMPLETE'
+    const projectId = resolvedVars.projectId;
+    const [albService] = await this.mustFindServices(['albService']);
+    const awsAccountId = await albService.findAwsAccountId(requestContext, projectId);
+    if (dependencyType === 'ALB') {
```
Can `dependencyType` be something else too? What is the rationale behind introducing `DEPENDENCY_TYPE`.

manikandan-thangavelu-rl replied:
This was based on a previous assumption. Have removed this logic and updated the PR

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js:
```diff
@@ -64,370 +64,386 @@ const successStatuses = ['CREATE_COMPLETE', 'DELETE_COMPLETE', 'UPDATE_COMPLETE'
  * Dependency check is performed based on the output of the product CFT
  */
 class CheckLaunchDependency extends StepBase {
```
Will you be adding tests for this class?

manikandan-thangavelu-rl replied:
Yes. You can expect unit test cases in the subsequent PRs

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/terminate-launch-dependency/terminate-launch-dependency.js:
```diff
@@ -0,0 +1,446 @@
+    };
+  }
+
+  async start() {
```
Would be helpful to add some info logging in this method around deleting records, rules and decreasing workspace count etc.

manikandan-thangavelu-rl replied:
Made the change and updated the PR

jn1119 commented on addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/terminate-launch-dependency/terminate-launch-dependency.js:
```diff
@@ -0,0 +1,446 @@
+ * Workflow step that terminates the launch dependencies of a product created while lauching the product
+ * Dependency check is performed based on the output of the product CFT
+ */
+class TerminateLaunchDependency extends StepBase {
```
I am guessing unit tests are pending for this class?

manikandan-thangavelu-rl replied:
Yes. You can expect unit test cases in the subsequent PRs

manikandan-thangavelu-rl committed:
Fix: Incorporated PR comments

RLIndia committed:
Merge pull request #15 from manikandan-thangavelu-rl/feat-rstudio-alb

Fix: Incorporated PR comments

jn1119 committed:
chore: Enable access logging for env-type-configs bucket (#520)

jn1119 committed:
chore: Enable server side encryption on prepare master and edge lambda bucket (#521)

jn1119 committed:
fix: Corrected Spark defaults to fix read/write functionality from Spark (#526)

jn1119 committed:
chore(release): 3.2.0 (#531)

dependabot[bot] committed:
chore(deps): bump glob-parent from 5.1.1 to 5.1.2 in /scripts/load-test-workspaces (#532)

* chore(deps): bump glob-parent in /scripts/load-test-workspaces

Bumps [glob-parent](https://github.com/gulpjs/glob-parent) from 5.1.1 to 5.1.2.
- [Release notes](https://github.com/gulpjs/glob-parent/releases)
- [Changelog](https://github.com/gulpjs/glob-parent/blob/main/CHANGELOG.md)
- [Commits](https://github.com/gulpjs/glob-parent/compare/v5.1.1...v5.1.2)

---
updated-dependencies:
- dependency-name: glob-parent
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update package-lock.json

* Update package-lock.json

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jeet <68876606+jn1119@users.noreply.github.com>

manikandan-thangavelu-rl committed:
Merge branch 'feat-rstudio-alb' into feat-rstudio-alb
